src/compose/app.ts

```diff
@@ -132,41 +136,62 @@ export class App {
 			true,
 		);
 
-		const { removePairs, installPairs, updatePairs } = this.compareServices(
-			this.services,
-			target.services,
-			state.containerIds,
-		);
+		const { removePairs, installPairs, updatePairs, dependentServices } =
+			this.compareServices(
+				this.services,
+				target.services,
+				state.containerIds,
+				networkChanges,
+				volumeChanges,
+			);
 
 		// For every service which needs to be updated, update via update strategy.
 		const servicePairs = removePairs.concat(updatePairs, installPairs);
-		steps = steps.concat(
-			servicePairs
-				.map((pair) =>
-					this.generateStepsForService(pair, {
-						...state,
-						servicePairs,
-						targetApp: target,
-						networkPairs: networkChanges,
-						volumePairs: volumeChanges,
-					}),
-				)
-				.filter((step) => step != null) as CompositionStep[],
-		);
+		// generateStepsForService will populate appsToLock with services that
+		// need to be locked, including services that need to be removed due to
+		// network or volume changes.
+		const appsToLock: AppsToLockMap = {
+			[this.appId]: new Set<string>(),
+			[target.appId]: new Set<string>(),
```

Comment:
If there's never a case where appIds would be different between this & target, this will be safe to remove. I was afraid of there being some case I may have missed even if I couldn't see it in the codebase, so I left this in

Reply:
As I mentioned, if moving between environment, `appUuid` for both apps could be the same but with different `appId` in which case this would make sense, although I suspect a bunch of other things would not work properly there. I also think our whole locking system presumes that apps are identified by `appId` so I think we can safely assume that both current and target apps have the same a`appId` for now

Reply:
I changed this [here](https://github.com/balena-os/balena-supervisor/pull/2234/commits/f199319d0a2ecf3dcbfc59c82f022b662d784a71#diff-857bb737cdeb3c33e2caa352769d3f96e0e8a48da6fea254eb99bc627732cec1R153-R156) to only construct a set where the key is target.appId